### PR TITLE
WIP: Add DEM tools

### PR DESCRIPTION
**Pull request by awirb**
Add DEM tools to generate idealised/generic topographies

### CODE REVIEW

File: avaframe/in3Utils/generateTopo.py [hunk 1]
```diff
@@ -0,0 +1,556 @@
...
+
+"""
+
+
+# load modules
+import numpy as np
+import matplotlib.pyplot as plt
+from mpl_toolkits.mplot3d import Axes3D
+from matplotlib import cm
+from scipy.stats import norm
```
Comment (fso42) [inline avaframe/in3Utils/generateTopo.py:15]: scipy needs to be added to the requirements.txt file

File: avaframe/in3Utils/generateTopo.py [hunk 2]
```diff
@@ -0,0 +1,561 @@
...
+        [x, y, z, name_ext] = hockeysmooth(cG, ncols, nrows, cT, cC, cF)
+
+    elif DEM_type == 'BL':
+        [x, y, z, name_ext] = bowl(cG, ncols, nrows, float(cT['r_bowl']))
+
+    elif DEM_type == 'HX':
+        [x, y, z, name_ext] = helix(cG, ncols, nrows, cT, f_len, A, B, cC, cF)
+
+    # Plot new topogrpahy
+    plotDEM(x, y, z, name_ext, cD, cG)
```
Comment (fso42) [inline avaframe/in3Utils/generateTopo.py:558]: I suggest plotting to file, or at least disable plottting with a flag -> right now it is blocking, meaning the user has to close the plot before program is allowed to continue...

File: avaframe/in3Utils/generateTopo.py [hunk 3]
```diff
@@ -0,0 +1,561 @@
...
+
+    # Write DEM to file
+    writeDEM(z, name_ext, ncols, nrows, float(cG['dx']), cD)
```
Comment (fso42) [inline avaframe/in3Utils/generateTopo.py:561]: see issue https://github.com/avaframe/AvaFrame/issues/36 , but no need to fix here

File: avaframe/in3Utils/generateTopoCfg.ini [hunk 4]
```diff
@@ -0,0 +1,80 @@
+### Config File - This file contains the main settings for the simulation run
+## Set your parameters
+# This file is part of Avaframe.
+
+
+# Domain parameters----------------------
+[GENERAL]
```
Comment (fso42) [inline avaframe/in3Utils/generateTopoCfg.ini:7]: See comments on generateTopo regarding combining sections

File: avaframe/in3Utils/generateTopoCfg.ini [hunk 5]
```diff
@@ -0,0 +1,80 @@
...
+[TOPO]
+f_lens = 2250
+mean_alpha = 24
+C = 1000
+r_bowl = 500
+r_helix = 1250
+z0 = 2200
+z_elev = 0
+r_circ = 200
+DEM_type = HS
```
Comment (fso42) [inline avaframe/in3Utils/generateTopoCfg.ini:27]: See comment in generateTopo.py regarding type:
List the possible options and their meaning here and in the docu...

File: avaframe/runGenerateTopo.py [hunk 6]
```diff
@@ -0,0 +1,17 @@
+"""
+    Run script for generateTopo in module in3Utils
+    This file is part of Avaframe.
+"""
+
+import sys
+import logging
+
+# Local imports
+import avaframe as af
```
Comment (fso42) [inline avaframe/runGenerateTopo.py:10]: Not used

File: docs/moduleIn3Utils.rst [hunk 7]
```diff
@@ -0,0 +1,78 @@
...
+Generate DEM files for idealised/generic topographies that can be used as required DEM input data for snow avalanche simulations.
+
+The following topography types can be chosen:
+
+* flat plane
+* inclined plane of constant slope
+* hockey stick - using a parabola to describe the sloping plane that then transitions into a flat foreland
+* hockey stick smooth - using an linearly sloping plane with smoothed transition to the flat foreland
+* bowl-shaped topography
+* helix-shaped topography
```
Comment (fso42) [inline docs/moduleIn3Utils.rst:13]: Add the short codes used in the ini file

File: docs/moduleIn3Utils.rst [hunk 8]
```diff
@@ -0,0 +1,78 @@
...
+
+* f_lens: 	distance to point where slope transitions into flat plane [m]
+* mean_alpha: 	slope angle from max. elevation to start flat plane [°] - or slope of inclined plane [°]
+* C: 		total fall height [m]
+* r_bowl:	  bowl radius [m]
+* r_helix:	  radius for helix [m]
+* z0: 		max elevation [m]
+* z_elev:	  elevation of flat plane [m]
+* r_circ: 	radius of smoothing circle [m]
+* DEM_type
```
Comment (fso42) [inline docs/moduleIn3Utils.rst:62]: See above

File: avaframe/in3Utils/generateTopo.py [hunk 9]
```diff
@@ -0,0 +1,556 @@
...
+
+  This file is part of Avaframe.
+
+"""
+
+
+# load modules
+import numpy as np
+import matplotlib.pyplot as plt
+from mpl_toolkits.mplot3d import Axes3D
```
[the remark this answers is not in the document]
Comment (awirb) [reply avaframe/in3Utils/generateTopo.py:13]: keep this line for the moment

File: avaframe/in3Utils/generateTopo.py [hunk 10]
```diff
@@ -0,0 +1,556 @@
...
+from matplotlib import cm
```
[the remark this answers is not in the document]
Comment (awirb) [reply avaframe/in3Utils/generateTopo.py:14]: Removed!

File: avaframe/in3Utils/generateTopo.py [hunk 11]
```diff
@@ -0,0 +1,556 @@
...
+from scipy.stats import norm
+import seaborn
```
[the remark this answers is not in the document]
Comment (awirb) [reply avaframe/in3Utils/generateTopo.py:16]: removed

File: avaframe/in3Utils/generateTopo.py [hunk 12]
```diff
@@ -0,0 +1,556 @@
...
+    mean_alpha = float(cT['mean_alpha'])
+    c_ff = float(cC['c_ff'])
+    c_radius = float(cC['c_radius'])
+
+    # Compute coordinate grid
+    xv = np.arange(0, x_end, dx)
+    yv = np.arange(-0.5*y_end, 0.5*y_end, dx)
+    x, y = np.meshgrid(xv, yv)
+    zv = np.zeros((nrows, ncols))
+    zv0 = np.zeros((nrows, ncols))
```
[the remark this answers is not in the document]
Comment (awirb) [reply avaframe/in3Utils/generateTopo.py:85]: zv0 removed

File: avaframe/in3Utils/generateTopo.py [hunk 13]
```diff
@@ -0,0 +1,556 @@
...
+    """
+        Compute coordinates of a parabolically-shaped slope with a flat foreland
+        defined by total fall height C, angle (mean_alpha) or distance (f_len) to flat foreland
+    """
+
+    # input parameters
+    dx = float(cG['dx'])
+    x_end = float(cG['x_end']) + dx
+    y_end = float(cG['y_end']) + dx
+    mean_alpha = float(cT['mean_alpha'])
```
[the remark this answers is not in the document]
Comment (awirb) [reply avaframe/in3Utils/generateTopo.py:263]: removed

File: avaframe/in3Utils/generateTopo.py [hunk 14]
```diff
@@ -0,0 +1,556 @@
...
+    cfg = configparser.ConfigParser()
+    if os.path.isfile('avaframe/in3Utils/local_generateTopoCfg.ini') == True:
+        cfg.read('avaframe/in3Utils/local_generateTopoCfg.ini')
+    else:
+        cfg.read('avaframe/in3Utils/generateTopoCfg.ini')
+    cG = cfg['GENERAL']
+    cT = cfg['TOPO']
+    cF = cfg['FLAGS']
+    cC = cfg['CHANNEL']
+    cD = cfg['DEMDATA']
```
[the remark this answers is not in the document]
Comment (awirb) [reply avaframe/in3Utils/generateTopo.py:522]: there is now cfgTopo (this is what was before cT, cG, CF)
cfgChannel
cfgDEM

File: avaframe/in3Utils/generateTopo.py [hunk 15]
```diff
@@ -0,0 +1,556 @@
...
+        [x, y, z, name_ext] = hockey(cG, f_len, A, B, ncols, nrows, cT, cC, cF)
+
+    elif DEM_type == 'HS2':
+        [x, y, z, name_ext] = hockeysmooth(cG, ncols, nrows, cT, cC, cF)
+
+    elif DEM_type == 'BL':
+        [x, y, z, name_ext] = bowl(cG, ncols, nrows, float(cT['r_bowl']))
+
+    elif DEM_type == 'HX':
+        [x, y, z, name_ext] = helix(cG, ncols, nrows, cT, f_len, A, B, cC, cF)
```
[the remark this answers is not in the document]
Comment (awirb) [reply avaframe/in3Utils/generateTopo.py:555]: Summarized cG, cT and cF into cfgTopo 
Added a description of the topo types in the .ini and the documentation

File: avaframe/in3Utils/generateTopo.py (hunk 1 above)
Comment (awirb) [reply avaframe/in3Utils/generateTopo.py:15]: scipy added in the requirements.txt file

File: avaframe/in3Utils/generateTopoCfg.ini (hunk 4 above)
Comment (awirb) [reply avaframe/in3Utils/generateTopoCfg.ini:7]: combined multiple sections see previous comments

File: avaframe/runGenerateTopo.py (hunk 6 above)
Comment (awirb) [reply avaframe/runGenerateTopo.py:10]: removed

File: docs/moduleIn3Utils.rst (hunk 7 above)
Comment (awirb) [reply docs/moduleIn3Utils.rst:13]: Added short codes

File: avaframe/in3Utils/generateTopo.py [hunk 16]
```diff
@@ -0,0 +1,556 @@
...
+import numpy as np
+import matplotlib.pyplot as plt
+from mpl_toolkits.mplot3d import Axes3D
+from matplotlib import cm
+from scipy.stats import norm
+import seaborn
+import os
+import configparser
+import logging
+logging.basicConfig(level=logging.INFO)
```
[the remark this answers is not in the document]
Comment (awirb) [reply avaframe/in3Utils/generateTopo.py:20]: Now local logger is created in generateTopo.py

File: avaframe/in3Utils/generateTopoCfg.ini (hunk 5 above)
Comment (awirb) [reply avaframe/in3Utils/generateTopoCfg.ini:27]: listed all topography types in the .ini file

File: avaframe/in3Utils/generateTopo.py [hunk 17]
```diff
@@ -0,0 +1,556 @@
...
+    # Compute distance to flat foreland for given mean_alpha
+    x1 = z0 / np.tan(np.radians(mean_alpha))
+    if x1 >= x_end * 0.9:
+        logging.warning('Your domain (x_end) is to small or the slope angle (mean_alpha) to shallow to produce a signifcant (>10 percent of domain, in your case: %.2f m) flat foreland!' % (0.1*(x_end-dx)))
+
+    # Compute circle parameters for smoothing the transition
+    beta = (0.5 * (180. - (mean_alpha)))
+    xc = r_circ / np.tan(np.radians(beta))
+    yc = xc * np.cos(np.radians(mean_alpha))
+    zc = r_circ
```
[the remark this answers is not in the document]
Comment (awirb) [reply avaframe/in3Utils/generateTopo.py:168]: this was for plotting the smoothing circle, but removed

File: avaframe/in3Utils/generateTopo.py [hunk 18]
```diff
@@ -0,0 +1,556 @@
...
+    # # Plot the smoothing transition
+    # plt.plot(xv, zv[1,:])
+    # plt.plot(xv, cv[1,:], 'r-')
+    # plt.plot(xv, lv, 'g-')
+    # plt.plot(x_circ, zc, 'g*')
+    # plt.title('Long profile of hockey stick smooth topo')
+    # plt.xlabel('Along valley distance [m]')
+    # plt.ylabel('Surface elevation [m]')
+    # plt.axis('equal')
+    # plt.show()
```
[the remark this answers is not in the document]
Comment (awirb) [reply avaframe/in3Utils/generateTopo.py:200]: removed

File: avaframe/in3Utils/generateTopo.py (hunk 2 above)
Comment (awirb) [reply avaframe/in3Utils/generateTopo.py:558]: plotting function is now moved to out3SimpPlot/outGenerateTopo.py and called in the runGenerateTopo.py file
plot is saved to file and can be shown when showplot flag is set to 1 in .ini file

File: docs/moduleIn3Utils.rst (hunk 8 above)
Comment (awirb) [reply docs/moduleIn3Utils.rst:62]: Added the abbreviations

**Comment by awirb** [comment]
Included the comments from review

**Pull request merged**